includes/class-wc-payments-webhook-processing-service.php

```diff
@@ -361,6 +361,15 @@ private function process_webhook_payment_intent_succeeded( $event_body ) {
 		$charges_data  = $this->read_webhook_property( $event_charges, 'data' );
 		$charge_id     = $this->read_webhook_property( $charges_data[0], 'id' );
 
+		if ( ! $order ) {
+			throw new Invalid_Webhook_Data_Exception(
+				sprintf(
+				/* translators: %1: charge ID */
+					__( 'Could not find order via charge ID: %1$s', 'woocommerce-payments' ),
+					$charge_id
+				)
+			);
+		}
```

Comment:
Looking into code it looks like the bug is caused by [this line](https://github.com/Automattic/woocommerce-payments/blob/5de214390589f707a8b0dc7472c680d06f8892c3/includes/class-wc-payments-webhook-processing-service.php#L577), according to it missing order is valid case and execution should continue, rather than throw exception. However, this change stops execution and throws the exception.

@htdat looks like you were the author of `get_order_from_event_body_intent_id` ([code](https://github.com/Automattic/woocommerce-payments/blob/5de214390589f707a8b0dc7472c680d06f8892c3/includes/class-wc-payments-webhook-processing-service.php#L559)), could you please clarify how the code is supposed to behave in fallback case.

Side note, this the second time when I see that the fatals are caused by mixed return type from the method.

Reply:
I think the bug comes from [this commit](https://github.com/Automattic/woocommerce-payments/commit/eb591d24146989485ce9ec65151e3f2e7bb4df95) where `$order` is used without checking it.

`get_order_from_event_body_intent_id` [was created here](https://github.com/Automattic/woocommerce-payments/commit/d619c4d5464899ba36cb187f8cfc7eebe7615238), but it just contains the code used in `process_webhook_payment_intent_succeeded` and was created to avoid duplicated code.

And as @vbelolapotkov mentions, we should not throw there, because that will prevent [`invoice.paid`](https://github.com/Automattic/woocommerce-payments/blob/5de214390589f707a8b0dc7472c680d06f8892c3/includes/class-wc-payments-webhook-processing-service.php#L169) to run.

Probably the simpler approach is to just check `$order` before using it in  `process_webhook_payment_intent_succeeded`. Although I don't know if that could bring any other issue. There are some new additions since I reviewed [the failed payments PR](https://github.com/Automattic/woocommerce-payments/pull/3628).

Reply:
Without Jessy PR, the code would break on a couple of lines bellows. I will create a PR that will resolve this problem. @ismaeldcom I think that the simplest way is to use your approach.

Reply:
Yes! `process_webhook_payment_intent_succeeded` was way shorter before, there were no additional uses of `$order`, but now we should wrap all of them within it or do an early return.

https://github.com/Automattic/woocommerce-payments/blob/d619c4d5464899ba36cb187f8cfc7eebe7615238/includes/admin/class-wc-rest-payments-webhook-controller.php#L287-L294

Reply:
> Side note, this the second time when I see that the fatals are caused by mixed return type from the method.

Either you return `false/null` or throw exception in this case. One of the problems is that I failed to see, as a reviewer, that this function can return multiple types. This should be covered with tests where every case that can occur during the execution of the function can occur.

Reply:
> I think the bug comes from [this commit](https://github.com/Automattic/woocommerce-payments/commit/eb591d24146989485ce9ec65151e3f2e7bb4df95) where `$order` is used without checking it.

Agreed this point! 

>`get_order_from_event_body_intent_id` [was created here](https://github.com/Automattic/woocommerce-payments/commit/d619c4d5464899ba36cb187f8cfc7eebe7615238), but it just contains the code used in `process_webhook_payment_intent_succeeded` and was created to avoid duplicated code.

But looking further, the side effect coming from this commit as it introduces `false` for this condition branch: 
https://github.com/Automattic/woocommerce-payments/blob/d619c4d5464899ba36cb187f8cfc7eebe7615238/includes/admin/class-wc-rest-payments-webhook-controller.php#L485-L488

The original commit [ref](https://github.com/Automattic/woocommerce-payments/commit/89766ab37e0b51af22497d8f14825a8b176528da) adding this condition means to return early and ignore this webhook:

https://github.com/Automattic/woocommerce-payments/blob/89766ab37e0b51af22497d8f14825a8b176528da/includes/admin/class-wc-rest-payments-webhook-controller.php#L273-L276

Reply:
> Probably the simpler approach is to just check `$order` before using it in `process_webhook_payment_intent_succeeded`.

Also agreed with this! And more specifically, we may want to return early for the `! $order` condition like the way we're handling in `process_webhook_payment_intent_failed`. 

```suggestion
		if ( ! $order ) {
			return;
		}
```

Reply:
>But looking further, the side effect coming from this commit as it introduces false for this condition branch:

I prefer null over false for those cases. But in the end it will throw an error if we don't check if `$order` is valid, right? 🤔

Reply:
@ismaeldcom - yes, returning `null` is ideal. But in this specific case, it just introduces another type for `get_order_from_event_body_intent_id` as in the body of this function, it's already having: `$this->wcpay_db->order_from_intent_id()` and `$this->wcpay_db->order_from_order_id()`, which are returning these types: `boolean|WC_Order|WC_Order_Refund`. 

For this specific issue, I think keeping the change is as light as possible, i.e. my suggestion above. For ideal return types, we may create an issue for it, and much better have a clear RFC/guideline to avoid this type of issue in the future.

Reply:
Sure! I was thinking on the first cause of the issue, but for fixing it now, I'd stick with an early check 🙂

Reply:
Added this commit https://github.com/Automattic/woocommerce-payments/pull/4287/commits/1c26bbd4e395e6cd197e7c989f42952073e06f03

I will take care of creating another issue later.

Reply:
> For this specific issue, I think keeping the change is as light as possible, i.e. my suggestion above. For ideal return types, we may create an issue for it, and much better have a clear RFC/guideline to avoid this kind of issue in the future.

FYI. Added the issue here https://github.com/Automattic/woocommerce-payments/issues/4293